firefox-ios/Client/Frontend/Browser/Toolbars/Redux/AddressBarState.swift

```diff
@@ -42,102 +42,119 @@ struct AddressBarState: StateType, Equatable {
 
         switch action.actionType {
         case ToolbarActionType.didLoadToolbars:
-            guard let model = (action as? ToolbarAction)?.addressToolbarModel else { return state }
-
-            return AddressBarState(
-                windowUUID: state.windowUUID,
-                navigationActions: model.navigationActions ?? state.navigationActions,
-                pageActions: model.pageActions ?? state.pageActions,
-                browserActions: model.browserActions ?? state.browserActions,
-                borderPosition: model.borderPosition ?? state.borderPosition,
-                url: model.url
-            )
+            return handleToolbarDidLoadToolbars(state: state, action: action)
 
         case ToolbarActionType.numberOfTabsChanged:
-            guard let numberOfTabs = (action as? ToolbarAction)?.numberOfTabs else { return state }
+            return handleToolbarNumberOfTabsChanged(state: state, action: action)
 
-            var actions = state.browserActions
+        case ToolbarActionType.urlDidChange:
+            return handleToolbarUrlDidChange(state: state, action: action)
 
-            if let index = actions.firstIndex(where: { $0.actionType == .tabs }) {
-                actions[index].numberOfTabs = numberOfTabs
-            }
+        case ToolbarActionType.backButtonStateChanged:
+            return handleToolbarBackButtonStateChanged(state: state, action: action)
 
-            return AddressBarState(
-                windowUUID: state.windowUUID,
-                navigationActions: state.navigationActions,
-                pageActions: state.pageActions,
-                browserActions: actions,
-                borderPosition: state.borderPosition,
-                url: state.url
-            )
+        case ToolbarActionType.forwardButtonStateChanged:
+            return handleToolbarForwardButtonStateChanged(state: state, action: action)
 
-        case ToolbarActionType.urlDidChange:
-            return AddressBarState(
-                windowUUID: state.windowUUID,
-                navigationActions: state.navigationActions,
-                pageActions: state.pageActions,
-                browserActions: state.browserActions,
-                borderPosition: state.borderPosition,
-                url: (action as? ToolbarAction)?.url
-            )
+        case ToolbarActionType.scrollOffsetChanged:
+            return handleToolbarScrollOffsetChanged(state: state, action: action)
 
-        case ToolbarActionType.backButtonStateChanged:
-            guard let isEnabled = (action as? ToolbarAction)?.isButtonEnabled else { return state }
+        default:
+            return state
```

Comment:
This is another case where the `struct` was recreated with identical values. I returned it directly because the outcome is the same. Does anyone have a different opinion?